caustic/parametrized.py

```diff
@@ -24,124 +25,84 @@ class Parametrized:
-        """
         return self._name
-
+    
     @name.setter
-    def name(self, newname: str):
-        """
-        Prevents the reassignment of the name attribute.
-
-        Raises:
-            NotImplementedError: Always, as reassigning the name attribute is not supported.
-        """
-        raise NotImplementedError()
-
-    def to(
-        self, device: Optional[torch.device] = None, dtype: Optional[torch.dtype] = None
-    ):
-        """
-        Moves static Params for this component and its descendants to the specified device and casts them to the specified data type.
-
-        Args:
-            device (Optional[torch.device], optional): The device to move the values to. Defaults to None.
-            dtype (Optional[torch.dtype], optional): The desired data type. Defaults to None.
-
-        Returns:
-            Parametrized: The Parametrized object itself after moving and casting.
-        """
-        for name in self._params:
+    def name(self, new_name: str):
+        old_name = self.name
+        for parent in self._parents.values():
+            parent._childs[new_name] = self
+            del parent._childs[old_name]
+        for child in self._childs.values():
+            child._parents[new_name] = self
+            del child._parents[old_name]
+        self._name = new_name
+
+    def to(self, device: Optional[torch.device] = None, dtype: Optional[torch.dtype] = None):
+        """
+        Moves static Params for this component and its childs to the specified device and casts them to the specified data type.
+        """
+        # TODO I think we should make users specify a separate method so we dont override this behavior
+        for name in self._params.keys():
             param = self._params[name]
             if isinstance(param, torch.Tensor):
                 self._params[name] = param.to(device, dtype)
-
-        for desc in self._descendants.values():
-            desc.to(device, dtype)
-
+        for child in self._childs.values():
+            child.to(device, dtype)
         return self
 
-    def _can_add(self, p: "Parametrized") -> bool:
-        """
-        Checks if a different model component with the same name is already in the DAG (Directed Acyclic Graph).
-
-        Args:
-            p (Parametrized): The Parametrized object to check.
-
-        Returns:
-            bool: False if a different model component with the same name is already in the DAG, True otherwise.
-        """
-        if self.name == p.name and self is not p:
-            return False
-
-        if p.name in self._descendants and self._descendants[p.name] is not p:
-            return False
-
-        # Propagate up through all parents
-        for parent in self._parents:
-            if not parent._can_add(p):
-                return False
-
-        return True
-
-    def _merge_descendants_up(self, p: "Parametrized"):
+    @staticmethod
+    def _generate_unique_name(name, module_names):
```

Comment:
Yes it might be an issue. Right now I assume that any module with the same name which do not have a direct edge to the current module are actually the same module. We will have such a situation when, for example, we do
```
class Sim(Simulator):
    def __init__(self):
        self.cosmo = FlatLambdaCDM()
        self.len = EPL(self.cosmo)
```

The DAG structure is
```

Sim -> FlatLambdaCDM
        -> EPL -> FlatLambdaCDM
```
We want to make sure FlatLambdaCDM is not renamed. Looking beyond the childs of Sim is a bit problematic and will cause FlatLambdaCDM to be tagged as a name collision when it is just a duplicate of the same noe. There might be another solution and mine certainly is not perfect. The code in place at least supports the most basic DAG structure (like the one above) with the correct behavior.